runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/QueryablePipeline.java

```diff
@@ -0,0 +1,277 @@
+        .nodes()
+        .stream()
+        .filter(pipelineNode -> pipelineNetwork.inEdges(pipelineNode).isEmpty())
+        .map(pipelineNode ->  (PTransformNode) pipelineNode)
+        .collect(Collectors.toSet());
+  }
+
+  public PTransformNode getProducer(PCollectionNode pcollection) {
+    return (PTransformNode) Iterables.getOnlyElement(pipelineNetwork.predecessors(pcollection));
+  }
+
+  /**
+   * Get all of the {@link PTransformNode PTransforms} which consume the provided {@link
+   * PCollectionNode} on a per-element basis.
+   *
+   * <p>If a {@link PTransformNode} consumes a {@link PCollectionNode} on a per-element basis one or
+   * more times, it will appear a single time in the result.
+   *
+   * <p>In theory, a transform may consume a single {@link PCollectionNode} in both a per-element
+   * and singleton manner. If this is the case, the transform node is included in the result, as it
+   * does consume the {@link PCollectionNode} on a per-element basis.
+   */
+  public Set<PTransformNode> getPerElementConsumers(PCollectionNode pCollection) {
+    return pipelineNetwork
+        .successors(pCollection)
+        .stream()
+        .filter(
+            consumer ->
+                pipelineNetwork
+                    .edgesConnecting(pCollection, consumer)
+                    .stream()
+                    .anyMatch(PipelineEdge::isPerElement))
+        .map(pipelineNode -> (PTransformNode) pipelineNode)
+        .collect(Collectors.toSet());
+  }
+
+  public Set<PCollectionNode> getProducedPCollections(PTransformNode ptransform) {
+    return pipelineNetwork
+        .successors(ptransform)
+        .stream()
+        .map(pipelineNode -> (PCollectionNode) pipelineNode)
+        .collect(Collectors.toSet());
+  }
+
+  public Components getComponents() {
+    return components;
+  }
+
+  /**
+   * Returns the {@link PCollectionNode PCollectionNodes} that the provided transform consumes as
+   * side inputs.
+   */
+  public Collection<PCollectionNode> getSideInputs(PTransformNode transform) {
+    return getLocalSideInputNames(transform.getTransform())
+        .stream()
+        .map(localName -> pCollectionNode(transform.getTransform().getInputsOrThrow(localName)))
+        .collect(Collectors.toSet());
+  }
+
+  private Set<String> getLocalSideInputNames(PTransform transform) {
+    if (PTransformTranslation.PAR_DO_TRANSFORM_URN.equals(transform.getSpec().getUrn())) {
+      try {
+        return ParDoPayload.parseFrom(transform.getSpec().getPayload()).getSideInputsMap().keySet();
+      } catch (InvalidProtocolBufferException e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      return Collections.emptySet();
+    }
+  }
+
+  /**
+   * Create the {@link PTransformNode} for the {@link PTransform} with the given ID.
+   *
+   * @throws IllegalArgumentException if there is no {@link PTransformNode} with the given ID.
+   */
+  public PTransformNode transformNode(String transformId) {
```

Comment:
Why would you want to expose this method and the pCollectionNode method below? Seems like this isn't a great shortcut compared to just calling PipelineNode...

Also, naming wise pTransformNode.

Reply:
Minor testing convenience, but inlined is NBD. Done.